### PR TITLE
Permission for player tp only

### PR DESCRIPTION
**Pull request by Emibergo02**
permission for only player tp
Proposed by Boy [here](https://discord.com/channels/818135932103557162/818531565662502933/1207281097579175996)

### CODE REVIEW

**Comment by WiIIiam278** [comment]
Could you also update the docs?

File: docs/Commands.md [hunk 1]
```diff
@@ -215,25 +215,25 @@ This is a table of HuskHomes commands, how to use them, and their required permi
...
-            <td><code>huskhomes.command.tp.coordinates</code></td>
+            <td><code>huskhomes.command.tp</code><br/><code>huskhomes.command.tp.coordinates</code></td>
             <td align="center">❌</td>
         </tr>
         <tr>
             <td><code>/tp &lt;teleporter_name&gt; &lt;coordinates&gt;</code></td>
             <td>Teleport an online user to a set of coordinates</td>
-            <td><code>huskhomes.command.tp.other</code><br/><code>huskhomes.command.tp.coordinates</code></td>
+            <td><code>huskhomes.command.tp</code><br/><code>huskhomes.command.tp.other</code><br/><code>huskhomes.command.tp.coordinates</code></td>
             <td align="center">❌</td>
```
Comment (WiIIiam278) [inline docs/Commands.md:237]: This is actually unnecessary and can be reverted. I based my initial comment on the fact you introduced a new permission node ("player").

File: common/src/main/java/net/william278/huskhomes/command/TpCommand.java [hunk 2]
```diff
@@ -140,7 +159,8 @@ public final List<String> suggest(@NotNull CommandUser user, @NotNull String[] a
                                     ((int) relative.getX() + " " + (int) relative.getY() + " " + (int) relative.getZ()))
                                     : List.of()
                     );
-                    completions.addAll(plugin.getPlayerList(false));
+                    if (servePlayerCompletions)
+                        completions.addAll(plugin.getPlayerList(false));
```
Comment (WiIIiam278) [inline common/src/main/java/net/william278/huskhomes/command/TpCommand.java:163]:
```suggestion
                    if (servePlayerCompletions) {
                        completions.addAll(plugin.getPlayerList(false));
                    }
```

File: common/src/main/java/net/william278/huskhomes/command/TpCommand.java [hunk 3]
```diff
@@ -121,7 +139,8 @@ public final List<String> suggest(@NotNull CommandUser user, @NotNull String[] a
                         ((int) relative.getX() + " " + (int) relative.getY()),
                         ((int) relative.getX() + " " + (int) relative.getY() + " " + (int) relative.getZ()))
                         : List.of());
-                completions.addAll(plugin.getPlayerList(false));
+                if (servePlayerCompletions)
+                    completions.addAll(plugin.getPlayerList(false));
```
Comment (WiIIiam278) [inline common/src/main/java/net/william278/huskhomes/command/TpCommand.java:143]:
```suggestion
                    if (servePlayerCompletions) {
                        completions.addAll(plugin.getPlayerList(false));
                    }
```

File: common/src/main/java/net/william278/huskhomes/command/TpCommand.java [hunk 4]
```diff
@@ -68,6 +79,12 @@ public void execute(@NotNull CommandUser executor, @NotNull String[] args) {
                     return;
                 }
 
+                //tp <teleporter_name> <coordinates> needs player AND coordinates permissions
```
Comment (WiIIiam278) [inline common/src/main/java/net/william278/huskhomes/command/TpCommand.java:82]:
```suggestion
                // tp <teleporter_name> <coordinates> requires both the .player and .coordinates permissions
```

File: common/src/main/java/net/william278/huskhomes/command/TpCommand.java [hunk 5]
```diff
@@ -49,10 +49,21 @@ public void execute(@NotNull CommandUser executor, @NotNull String[] args) {
...
             }
-            case 2 -> this.execute(executor, Teleportable.username(args[0]), Target.username(args[1]), args);
+            case 2 -> {
+                if (!executor.hasPermission(getPermission("other"))) {
+                    plugin.getLocales().getLocale("error_no_permission")
+                            .ifPresent(executor::sendMessage);
+                    return;
+                }
+                this.execute(executor, Teleportable.username(args[0]), Target.username(args[1]), args);
+            }
```
Comment (WiIIiam278) [inline common/src/main/java/net/william278/huskhomes/command/TpCommand.java:66]: Okay - but what if the player passes their own username as `<username>`? We shouldn't require this permission node for the player to teleport themselves.

We should move these checks to the `execute` logic - checking if 
* the person being teleported is the same as the person executing the teleport
* and, if not, they have this permission node there.

**Review by WiIIiam278 — CHANGES_REQUESTED** [review]
Needs some changes

File: common/src/main/java/net/william278/huskhomes/command/TpCommand.java [hunk 6]
```diff
@@ -88,9 +87,23 @@ private void execute(@NotNull CommandUser executor, @NotNull Teleportable telepo
         final TeleportBuilder builder = Teleport.builder(plugin)
                 .teleporter(teleportable)
                 .target(target);
-        if (executor instanceof OnlineUser user) {
-            builder.executor(user);
+
+        if (executor instanceof OnlineUser onlineUser) {
+            if (!onlineUser.hasPermission(getPermission("other"))) {
```
Comment (WiIIiam278) [inline common/src/main/java/net/william278/huskhomes/command/TpCommand.java:92]: Yeah, but what if the `executor` _is_ the person being teleported? We shouldn't require "other" in this case.

Comment (Emibergo02) [reply common/src/main/java/net/william278/huskhomes/command/TpCommand.java]: I moved it into the teleportable if clause

File: common/src/main/java/net/william278/huskhomes/command/TpCommand.java [hunk 7]
```diff
@@ -88,9 +87,24 @@ private void execute(@NotNull CommandUser executor, @NotNull Teleportable telepo
         final TeleportBuilder builder = Teleport.builder(plugin)
                 .teleporter(teleportable)
                 .target(target);
-        if (executor instanceof OnlineUser user) {
-            builder.executor(user);
+
+        if (executor instanceof OnlineUser onlineUser) {
+            if (target instanceof Teleportable teleportableTarget) {
+                if (!onlineUser.hasPermission(getPermission("other"))) {
```
Comment (WiIIiam278) [inline common/src/main/java/net/william278/huskhomes/command/TpCommand.java:93]: This logic is _still_ wrong - if `onlineUser` and `teleportableTarget` are the same person, this check will still incorrectly be triggered.

**Review by WiIIiam278 — APPROVED** [review]

**Pull request merged**